File: src/main/java/ltdjms/discord/shared/cache/NoOpCacheService.java
package ltdjms.discord.shared.cache;

import java.util.Optional;

/** No-op 實現的 CacheService，用於測試或禁用緩存的場景。 所有操作都不執行任何動作，get 方法始終返回 Optional.empty()。 */
public class NoOpCacheService implements CacheService {

  private static final NoOpCacheService INSTANCE = new NoOpCacheService();

  private NoOpCacheService() {
    // Private constructor for singleton
  }

  /**
   * 獲取 NoOpCacheService 的單例實例。
   *
   * @return NoOpCacheService 實例
   */
  public static CacheService getInstance() {
    return INSTANCE;
  }

  @Override
  public <T> Optional<T> get(String key, Class<T> type) {
    return Optional.empty();
  }

  @Override
  public <T> void put(String key, T value, int ttlSeconds) {
    // No-op
  }

  @Override
  public void invalidate(String key) {
    // No-op
  }
}

File: packages/shared/src/infra/cache/noop-cache-service.ts
import { type CacheService } from './cache-service.js';

/**
 * No-op implementation of CacheService for testing or when caching is disabled.
 * get always returns null, put/invalidate do nothing.
 * Matches Java NoOpCacheService.
 */
export class NoOpCacheService implements CacheService {
  private static readonly INSTANCE = new NoOpCacheService();

  private constructor() {}

  static getInstance(): CacheService {
    return NoOpCacheService.INSTANCE;
  }

  async get<T>(_key: string): Promise<T | null> {
    return null;
  }

  async put(_key: string, _value: unknown, _ttlSeconds: number): Promise<void> {
    // No-op
  }

  async invalidate(_key: string): Promise<void> {
    // No-op
  }

  async exists(_key: string): Promise<boolean> {
    return false;
  }
}
